Zed crashes when edition JSON file over SSH

### Summary

<!-- Begin your issue with a one sentence summary -->
Zed crashes upon connecting to remote host

### Description
<!--  Describe with sufficient detail to reproduce from a clean Zed install.
  - Any code must be sufficient to reproduce (include context!)
  - Code must as text, not just as a screenshot.
  - Issues with insufficient detail may be summarily closed.
-->

On the latest Zed Preview, Zed immediately loses connection to any remote host upon editing a JSON file, even with a blank `settings.json` file.

Viewing the logs for the remote server shows the following stacktrace:

```
(remote server) panic occurred: called `Result::unwrap()` on an `Err` value: no state of type theme::registry::GlobalThemeRegistry exists
Backtrace:
remote_server::unix::init_panic_hook::{{closure}}
std::panicking::rust_panic_with_hook
std::panicking::begin_panic_handler::{{closure}}
std::sys::backtrace::__rust_end_short_backtrace
__rustc::rust_begin_unwind
core::panicking::panic_fmt
core::result::unwrap_failed
theme::registry::ThemeRegistry::global
core::ops::function::FnOnce::call_once
settings::settings_store::SettingsStore::json_schema
languages::json::JsonLspAdapter::get_workspace_config
<languages::json::JsonLspAdapter as language::LspAdapter>::workspace_configuration::{{closure}}project::lsp_store::LocalLspStore::workspace_configuration_for_adapter::{{closure}}
project::lsp_store::LocalLspStore::start_language_server::{{closure}}::{{closure}}
gpui::app::App::spawn::{{closure}}
async_task::raw::RawTask<F,T,S,M>::run
<core::cell::RefCell<calloop::sources::DispatcherInner<S,F>> as calloop::sources::EventDispatcher<Data>>::process_events
<gpui::platform::linux::headless::client::HeadlessClient as gpui::platform::linux::platform::LinuxClient>::run
gpui::platform::linux::platform::<impl gpui::platform::Platform for P>::run
remote_server::unix::execute_run
remote_server::main
std::sys::backtrace::__rust_begin_short_backtrace
std::rt::lang_start::{{closure}}
std::rt::lang_start_internal
main
```

which seems to suggest the theme registry isn't being initialized properly?


Steps to reproduce:
1. Install the latest Zed Preview (Zed Preview 0.194.0 
98b8692fde48a33616300d9ae40386c8ba4c92d1)
2. Connect to any remote host.
3. Open any JSON file. I used `{"foo": "bar"}
4. Zed immediately loses connection.

**Expected Behavior**:
Zed is able to remain connected to a remote host.
**Actual Behavior**:
Zed immediately loses the connection.
<!-- Before Submitting, did you:
  1. Include settings.json, keymap.json, .editorconfig if relevant?
  2. Check your Zed.log for relevant errors? (please include!)
  3. Click Preview to ensure everything looks right?
  4. Hide videos, large images and logs in ``` inside collapsible blocks:

<details><summary>click to expand</summary>

</details>


### Zed Version and System Specs

Zed: v0.194.0 (Zed Preview) 
OS: macOS 15.5.0
Memory: 64 GiB
Architecture: aarch64

I am seeing something similar on 0.194.0 with a react file. possibly eslint related?

```
2025-07-02T15:29:19-07:00 ERROR [gpui] window not found
2025-07-02T15:29:27-07:00 ERROR [remote::ssh_session] client:unhandled ssh message name:UpdateProject
2025-07-02T15:29:33-07:00 ERROR [lsp] (remote server) cannot read LSP message headers
2025-07-02T15:29:33-07:00 ERROR [lsp] (remote server) Server reset connection for a request "initialize" id 0
2025-07-02T15:29:33-07:00 ERROR [project::lsp_store] (remote server) Failed to start language server "eslint": Error {
    context: "initializing server eslint, id 1",
    source: "Server reset the connection",
}
2025-07-02T15:29:33-07:00 ERROR [project::lsp_store] (remote server) server stderr: node:internal/modules/cjs/loader:1222
  throw err;
  ^

Error: Cannot find module '/home/.local/share/zed/languages/eslint/vscode-eslint-2.4.4/vscode-eslint/server/out/eslintServer.js'
    at Module._resolveFilename (node:internal/modules/cjs/loader:1219:15)
    at Module._load (node:internal/modules/cjs/loader:1045:27)
    at TracingChannel.traceSync (node:diagnostics_channel:315:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:215:24)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:158:5)
    at node:internal/main/run_main_module:30:49 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}

Node.js v22.5.1

2025-07-02T15:29:33-07:00 ERROR [lsp] (remote server) server shut down
2025-07-02T15:29:35-07:00 ERROR [remote_server::unix] (remote proxy) encountered error while forwarding messages: stdout_task failed

Caused by:
    0: failed to read message from stdout
    1: unexpected end of file, terminating...
2025-07-02T15:29:35-07:00 ERROR [remote_server] (remote proxy) exiting due to error: stdout_task failed
2025-07-02T15:29:35-07:00 WARN  [remote::ssh_session] ssh heartbeat: connection activity channel has been dropped. stopping.
2025-07-02T15:29:35-07:00 ERROR [remote::ssh_session] proxy process terminated unexpectedly
2025-07-02T15:29:35-07:00 INFO  [remote::ssh_session] setting state to 'reconnecting'
2025-07-02T15:29:35-07:00 INFO  [remote::ssh_session] Trying to reconnect to ssh server... Attempt 1
2025-07-02T15:29:38-07:00 ERROR [remote_server::unix] (remote proxy) attempted to reconnect, but no server running
2025-07-02T15:29:38-07:00 ERROR [remote::ssh_session] failed to reconnect because server is not running
2025-07-02T15:29:38-07:00 INFO  [remote::ssh_session] setting state to 'server not running'
2025-07-02T15:29:40-07:00 ERROR [remote::ssh_session] client:unhandled ssh message name:UpdateProject
2025-07-02T15:29:40-07:00 INFO  [remote::ssh_session] shutting down ssh processes
2025-07-02T15:29:43-07:00 ERROR [remote] entity released
```

```
Language server eslint:

initializing server eslint, id 2
-- stderr--
node:internal/modules/cjs/loader:1222
  throw err;
  ^

Error: Cannot find module '/home/bento/.local/share/zed/languages/eslint/vscode-eslint-2.4.4/vscode-eslint/server/out/eslintServer.js'
    at Module._resolveFilename (node:internal/modules/cjs/loader:1219:15)
    at Module._load (node:internal/modules/cjs/loader:1045:27)
    at TracingChannel.traceSync (node:diagnostics_channel:315:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:215:24)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:158:5)
    at node:internal/main/run_main_module:30:49 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}

Node.js v22.5.1
```

--- a/crates/language/src/language_settings.rs
+++ b/crates/language/src/language_settings.rs
@@ -321,7 +321,7 @@ inventory::submit! {
             let language_settings_content_ref = generator
                 .subschema_for::<LanguageSettingsContent>()
                 .to_value();
-            let schema = json_schema!({
+            replace_subschema::<LanguageToSettingsMap>(generator, || json_schema!({
                 "type": "object",
                 "properties": params
                     .language_names
@@ -333,8 +333,7 @@ inventory::submit! {
                         )
                     })
                     .collect::<serde_json::Map<_, _>>()
-            });
-            replace_subschema::<LanguageToSettingsMap>(generator, schema)
+            }))
         }
     }
 }

--- a/crates/settings/src/settings_json.rs
+++ b/crates/settings/src/settings_json.rs
@@ -23,35 +23,26 @@ inventory::collect!(ParameterizedJsonSchema);
 
 const DEFS_PATH: &str = "#/$defs/";
 
-/// Replaces the JSON schema definition for some type, and returns a reference to it.
+/// Replaces the JSON schema definition for some type if it is in use (in the definitions list), and
+/// returns a reference to it.
+///
+/// This asserts that JsonSchema::schema_name() + "2" does not exist because this indicates that
+/// there are multiple types that use this name, and unfotunately schemars APIs do not support
+/// resolving this ambiguity - see https://github.com/GREsau/schemars/issues/449
+///
+/// This takes a closure for `schema` because some settings types are not available on the remote
+/// server, and so will crash when attempting to access e.g. GlobalThemeRegistry.
 pub fn replace_subschema<T: JsonSchema>(
     generator: &mut schemars::SchemaGenerator,
-    schema: schemars::Schema,
+    schema: impl Fn() -> schemars::Schema,
 ) -> schemars::Schema {
-    // The key in definitions may not match T::schema_name() if multiple types have the same name.
-    // This is a workaround for there being no straightforward way to get the key used for a type -
-    // see https://github.com/GREsau/schemars/issues/449
-    let ref_schema = generator.subschema_for::<T>();
-    if let Some(serde_json::Value::String(definition_pointer)) = ref_schema.get("$ref") {
-        if let Some(definition_name) = definition_pointer.strip_prefix(DEFS_PATH) {
-            generator
-                .definitions_mut()
-                .insert(definition_name.to_string(), schema.to_value());
-            return ref_schema;
-        } else {
-            log::error!(
-                "bug: expected `$ref` field to start with {DEFS_PATH}, \
-                got {definition_pointer}"
-            );
-        }
-    } else {
-        log::error!("bug: expected `$ref` field in result of `subschema_for`");
-    }
     // fallback on just using the schema name, which could collide.
     let schema_name = T::schema_name();
-    generator
-        .definitions_mut()
-        .insert(schema_name.to_string(), schema.to_value());
+    let definitions = generator.definitions_mut();
+    assert!(!definitions.contains_key(&format!("{schema_name}2")));
+    if definitions.contains_key(schema_name.as_ref()) {
+        definitions.insert(schema_name.to_string(), schema().to_value());
+    }
     Schema::new_ref(format!("{DEFS_PATH}{schema_name}"))
 }
 

--- a/crates/settings/src/settings_json.rs
+++ b/crates/settings/src/settings_json.rs
@@ -27,7 +27,7 @@ const DEFS_PATH: &str = "#/$defs/";
 /// returns a reference to it.
 ///
 /// This asserts that JsonSchema::schema_name() + "2" does not exist because this indicates that
-/// there are multiple types that use this name, and unfotunately schemars APIs do not support
+/// there are multiple types that use this name, and unfortunately schemars APIs do not support
 /// resolving this ambiguity - see https://github.com/GREsau/schemars/issues/449
 ///
 /// This takes a closure for `schema` because some settings types are not available on the remote

--- a/crates/theme/src/settings.rs
+++ b/crates/theme/src/settings.rs
@@ -978,11 +978,10 @@ pub struct ThemeName(pub Arc<str>);
 inventory::submit! {
     ParameterizedJsonSchema {
         add_and_get_ref: |generator, _params, cx| {
-            let schema = json_schema!({
+            replace_subschema::<ThemeName>(generator, || json_schema!({
                 "type": "string",
                 "enum": ThemeRegistry::global(cx).list_names(),
-            });
-            replace_subschema::<ThemeName>(generator, schema)
+            }))
         }
     }
 }
@@ -996,15 +995,14 @@ pub struct IconThemeName(pub Arc<str>);
 inventory::submit! {
     ParameterizedJsonSchema {
         add_and_get_ref: |generator, _params, cx| {
-            let schema = json_schema!({
+            replace_subschema::<IconThemeName>(generator, || json_schema!({
                 "type": "string",
                 "enum": ThemeRegistry::global(cx)
                     .list_icon_themes()
                     .into_iter()
                     .map(|icon_theme| icon_theme.name)
                     .collect::<Vec<SharedString>>(),
-            });
-            replace_subschema::<IconThemeName>(generator, schema)
+            }))
         }
     }
 }
@@ -1018,11 +1016,12 @@ pub struct FontFamilyName(pub Arc<str>);
 inventory::submit! {
     ParameterizedJsonSchema {
         add_and_get_ref: |generator, params, _cx| {
-            let schema = json_schema!({
-                "type": "string",
-                "enum": params.font_names,
-            });
-            replace_subschema::<FontFamilyName>(generator, schema)
+            replace_subschema::<FontFamilyName>(generator, || {
+                json_schema!({
+                    "type": "string",
+                    "enum": params.font_names,
+                })
+            })
         }
     }
 }

